src/muya/lib/parser/utils.js

```diff
@@ -0,0 +1,104 @@
+// These utils only used for this folder.
+// Move punctuation and WHITELIST_ATTRIBUTES here just to prevent `navigator is not defined` error when run test.
+export const punctuation = ['!', '"', '#', '$', '%', '&', "'", '(', ')', '*', '+', ',', '-', '.', '/', ':', ';', '<', '=', '>', '?', '@', '[', '\\', ']', '^', '_', '`', '{', '|', '}', '~']
+// selected from https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes
+export const WHITELIST_ATTRIBUTES = [
+  'align', 'alt', 'checked', 'class', 'color', 'dir', 'disabled', 'for', 'height', 'hidden',
+  'href', 'id', 'lang', 'lazyload', 'rel', 'spellcheck', 'src', 'srcset', 'start', 'style',
+  'target', 'title', 'type', 'value', 'width'
+]
+
+const validWidthAndHeight = value => {
+  if (!/^\d{1,}$/.test(value)) return ''
+  value = parseInt(value)
+  return value >= 0 ? value : ''
+}
+
+export const lowerPriority = (src, offset, rules) => {
+  let i
+  for (i = 0; i < offset; i++) {
+    const text = src.substring(i)
+    for (const rule of Object.keys(rules)) {
+      const to = rules[rule].exec(text)
+      if (to && to[0].length > offset - i) {
+        return false
+      }
+    }
+  }
+  return true
+}
+
+export const getAttributes = html => {
+  const parser = new DOMParser()
+  const doc = parser.parseFromString(html, 'text/html')
+  const target = doc.querySelector('body').firstElementChild
+  if (!target) return null
+  const attrs = {}
+  for (const attr of target.getAttributeNames()) {
+    if (!WHITELIST_ATTRIBUTES.includes(attr)) continue
+    if (/width|height/.test(attr)) {
+      attrs[attr] = validWidthAndHeight(target.getAttribute(attr))
+    } else {
+      attrs[attr] = target.getAttribute(attr)
+    }
+  }
+
+  return attrs
+}
+
+export const parseSrcAndTitle = (text = '') => {
+  const parts = text.split(/\s+/)
+  const src = parts[0]
+  const rawTitle = text.substring(src.length).trim()
+  const TITLE_REG = /^('|")(.*?)\1$/ // we only support use `'` and `"` to indicate a title now.
+  let title = ''
+  if (rawTitle && TITLE_REG.test(rawTitle)) {
+    title = rawTitle.replace(TITLE_REG, '$2')
+  }
+  return { src, title }
+}
+
+export const validateEmphasize = (src, offset, marker, pending, rules) => {
+  /**
+   * Intraword emphasis is disallowed for _
+   */
+  const lastChar = pending.charAt(pending.length - 1) || ''
+  const followedChar = src[offset]
+  const ALPHA_REG = /[a-zA-Z]{1}/
```

Comment:
You should use a Regex like [this](https://github.com/fxha/marktext/blob/07534df1be75ad52f751f8a512809f33b59671b7/src/renderer/util/spellchecker.js#L13) ([example](https://github.com/fxha/marktext/blob/07534df1be75ad52f751f8a512809f33b59671b7/src/renderer/util/spellchecker.js#L242)) to parse text like example 374 correct.

Reply:
I followed the commonMark Spec, and write `canOpenEmphasis` and `canCloseEmphasis` method, so I think `ALPHA_REG` is not needed.